services/blockstorage/block_sync.go

```diff
@@ -0,0 +1,323 @@
+package blockstorage
+
+import (
+	"context"
+	"github.com/orbs-network/orbs-network-go/instrumentation/log"
+	"github.com/orbs-network/orbs-spec/types/go/primitives"
+	"github.com/orbs-network/orbs-spec/types/go/protocol"
+	"github.com/orbs-network/orbs-spec/types/go/protocol/gossipmessages"
+	"github.com/orbs-network/orbs-spec/types/go/services"
+	"github.com/orbs-network/orbs-spec/types/go/services/gossiptopics"
+	"github.com/pkg/errors"
+	"math/rand"
+	"time"
+)
+
+type blockSyncState int
+
+const (
+	BLOCK_SYNC_STATE_IDLE                                   blockSyncState = 0
+	BLOCK_SYNC_STATE_START_SYNC                             blockSyncState = 1
+	BLOCK_SYNC_PETITIONER_COLLECTING_AVAILABILITY_RESPONSES blockSyncState = 2
+	BLOCK_SYNC_PETITIONER_ASK_FOR_BLOCKS                    blockSyncState = 3
+	BLOCK_SYNC_PETITIONER_WAITING_FOR_CHUNK                 blockSyncState = 4
+)
+
+type BlockSyncStorage interface {
+	GetBlocks(first primitives.BlockHeight, last primitives.BlockHeight) (blocks []*protocol.BlockPairContainer, firstAvailableBlockHeight primitives.BlockHeight, lastAvailableBlockHeight primitives.BlockHeight)
+	LastCommittedBlockHeight() primitives.BlockHeight
+	CommitBlock(input *services.CommitBlockInput) (*services.CommitBlockOutput, error)
+	ValidateBlockForCommit(input *services.ValidateBlockForCommitInput) (*services.ValidateBlockForCommitOutput, error)
+	UpdateConsensusAlgo()
+}
+
+type BlockSync struct {
+	reporting log.BasicLogger
+
+	config  Config
+	storage BlockSyncStorage
+	gossip  gossiptopics.BlockSync
+	Events  chan interface{}
+}
+
+func NewBlockSync(ctx context.Context, config Config, storage BlockSyncStorage, gossip gossiptopics.BlockSync, reporting log.BasicLogger) *BlockSync {
+	blockSync := &BlockSync{
+		reporting: reporting.For(log.String("flow", "block-sync")),
+		config:    config,
+		storage:   storage,
+		gossip:    gossip,
+		Events:    make(chan interface{}),
+	}
+
+	go blockSync.mainLoop(ctx)
+
+	return blockSync
+}
+
+func (b *BlockSync) mainLoop(ctx context.Context) {
+	state := BLOCK_SYNC_STATE_START_SYNC
+	var event interface{}
+	var blockAvailabilityResponses []*gossipmessages.BlockAvailabilityResponseMessage
+
+	syncTrigger := time.AfterFunc(b.config.BlockSyncInterval(), func() {
+		if state == BLOCK_SYNC_STATE_IDLE {
+			state = BLOCK_SYNC_STATE_START_SYNC
+			b.Events <- nil
+		}
+	})
+
+	requestBlocksTrigger := time.AfterFunc(b.config.BlockSyncCollectResponseTimeout(), func() {
+		if state == BLOCK_SYNC_PETITIONER_COLLECTING_AVAILABILITY_RESPONSES {
+			state = BLOCK_SYNC_PETITIONER_ASK_FOR_BLOCKS
+			b.Events <- nil
+		}
+	})
+
+	for {
```

Comment:
We should extract the for body to a different function that receives the current state and returns the next state and handles the entire body.

This extracted function is the function that we will test in the unit tests directly.

Reply:
I don't think it's possible now because it requires another refactoring. I will do it in another pr.